File: backend/routers/accident_detection.py
import cv2
import numpy as np

from fastapi import (
    APIRouter,
    File,
    HTTPException,
    UploadFile
)

from services.accident_detection_service import (
    AccidentDetectionService
)


# ============================================================
# ROUTER
# ============================================================

router = APIRouter(

    prefix="/api/accident-detection",

    tags=["Emergency Accident Detection"]

)


# ============================================================
# LOAD ACCIDENT AI MODEL ONCE
# ============================================================
#
# IMPORTANT:
#
# The model is loaded when FastAPI starts.
#
# It is NOT loaded every time a frame is uploaded.
#
# ============================================================

detector = AccidentDetectionService(

    confidence_threshold=0.35,

    required_hits=3,

    max_missed_frames=10

)


# ============================================================
# ANALYZE LIVE CAMERA FRAME
# ============================================================

@router.post("/frame")
async def analyze_accident_frame(

    file: UploadFile = File(...)

):


    try:


        # ====================================================
        # VALIDATE FILE
        # ====================================================

        if not file.content_type:


            raise HTTPException(

                status_code=400,

                detail=(
                    "No valid image content type provided."
                )

            )


        if not file.content_type.startswith(

            "image/"

        ):


            raise HTTPException(

                status_code=400,

                detail=(
                    "Invalid file type. "
                    "Please upload an image frame."
                )

            )


        # ====================================================
        # READ IMAGE
        # ====================================================

        image_bytes = await file.read()


        if not image_bytes:


            raise HTTPException(

                status_code=400,

                detail="Uploaded image is empty."

            )


        # ====================================================
        # CONVERT BYTES → NUMPY
        # ====================================================

        image_array = np.frombuffer(

            image_bytes,

            dtype=np.uint8

        )


        # ====================================================
        # DECODE IMAGE
        # ====================================================

        frame = cv2.imdecode(

            image_array,

            cv2.IMREAD_COLOR

        )


        if frame is None:


            raise HTTPException(

                status_code=400,

                detail="Could not decode image frame."

            )


        # ====================================================
        # RUN ACCIDENT AI
        # ====================================================

        result = detector.analyze_frame(

            frame

        )


        # ====================================================
        # FORMAT DETECTIONS
        # ====================================================

        detections = []


        for detection in result[

            "raw_detections"

        ]:


            detections.append({

                "class_id":
                    detection["class_id"],


                "type":
                    detection["type"],


                "confidence":
                    round(

                        detection["confidence"],

                        4

                    ),


                "bounding_box":
                    detection["bounding_box"],


                "is_accident":
                    detection["is_accident"]

            })


        # ====================================================
        # RESPONSE
        # ====================================================

        return {


            "success":
                True,


            # ------------------------------------------------
            # RAW MODEL DETECTIONS
            # ------------------------------------------------

            "detections":
                detections,


            # ------------------------------------------------
            # ACCIDENT STATUS
            # ------------------------------------------------

            "accident_detected":
                result["accident_detected"],


            "accident_active":
                result["accident_active"],


            # True ONLY when accident becomes confirmed
            "newly_confirmed":
                result["newly_confirmed"],


            # ------------------------------------------------
            # TEMPORAL VERIFICATION
            # ------------------------------------------------

            "verification": {

                "consecutive_hits":
                    result["consecutive_hits"],


                "required_hits":
                    result["required_hits"],


                "missed_frames":
                    result["missed_frames"]

            },


            # ------------------------------------------------
            # BEST ACCIDENT DETECTION
            # ------------------------------------------------

            "best_accident":
                result["best_accident"],


            # ------------------------------------------------
            # STATISTICS
            # ------------------------------------------------

            "statistics": {

                "total_frames":
                    result["total_frames"],


                "confirmed_accidents":
                    result["confirmed_accidents"]

            },


            # ------------------------------------------------
            # FRAME INFORMATION
            # ------------------------------------------------

            "frame_width":
                frame.shape[1],


            "frame_height":
                frame.shape[0]

        }


    except HTTPException:


        raise


    except Exception as error:


        print(
            "\nACCIDENT DETECTION ERROR:"
        )


        print(
            str(error)
        )


        raise HTTPException(

            status_code=500,

            detail=(

                "Accident frame analysis failed: "

                f"{str(error)}"

            )

        )


# ============================================================
# ACCIDENT AI STATUS
# ============================================================

@router.get("/status")
def get_accident_ai_status():


    return {


        "status":
            "READY",


        "model_loaded":
            True,


        "model_path":
            detector.model_path,


        "confidence_threshold":
            detector.confidence_threshold,


        "required_hits":
            detector.required_hits,


        "max_missed_frames":
            detector.max_missed_frames,


        "model_classes":
            detector.model.names,


        "total_frames":
            detector.total_frames,


        "confirmed_accidents":
            detector.confirmed_accidents

    }


# ============================================================
# RESET ACCIDENT DETECTOR
# ============================================================

@router.post("/reset")
def reset_accident_detector():


    return detector.reset()

File: frontend/src/services/accidentDetection.ts
import { api } from './api';
// ============================================================
// ACCIDENT DETECTION TYPES
// ============================================================

export interface BoundingBox {

  x1: number;

  y1: number;

  x2: number;

  y2: number;

}


export interface AccidentDetection {

  class_id: number;

  type: string;

  confidence: number;

  bounding_box: BoundingBox;

  is_accident: boolean;

}


export interface AccidentVerification {

  consecutive_hits: number;

  required_hits: number;

  missed_frames: number;

}


export interface AccidentStatistics {

  total_frames: number;

  confirmed_accidents: number;

}


export interface AccidentFrameResponse {

  success: boolean;

  detections: AccidentDetection[];

  accident_detected: boolean;

  accident_active: boolean;

  newly_confirmed: boolean;

  verification: AccidentVerification;

  best_accident: AccidentDetection | null;

  statistics: AccidentStatistics;

  frame_width: number;

  frame_height: number;

}
// ============================================================
// IMAGE DATA → JPEG BLOB
// ============================================================

async function imageDataToBlob(

  frame: ImageData

): Promise<Blob> {


  const canvas =
    document.createElement('canvas');


  canvas.width =
    frame.width;


  canvas.height =
    frame.height;


  const context =
    canvas.getContext('2d');


  if (!context) {

    throw new Error(
      'Unable to create canvas context.'
    );

  }

  // Put captured frame into canvas

  context.putImageData(

    frame,

    0,

    0

  );


  // Convert to JPEG

  return new Promise(

    (

      resolve,

      reject

    ) => {

      canvas.toBlob(

        (blob) => {

          if (!blob) {

            reject(

              new Error(
                'Unable to convert frame to JPEG.'
              )

            );

            return;

          }


          resolve(
            blob
          );

        },


        'image/jpeg',


        0.90

      );

    }

  );

}
// ============================================================
// ANALYZE ACCIDENT FRAME
// ============================================================

export async function analyzeAccidentFrame(

  frame: ImageData

): Promise<AccidentFrameResponse> {


  try {


    // ========================================================
    // CONVERT FRAME → JPEG
    // ========================================================

    const imageBlob =
      await imageDataToBlob(
        frame
      );


    // ========================================================
    // CREATE FORM DATA
    // ========================================================

    const formData =
      new FormData();


    formData.append(

      'file',

      imageBlob,

      'accident-frame.jpg'

    );


    // ========================================================
    // SEND TO ACCIDENT AI BACKEND
    // ========================================================

    const { data } =
      await api.post(

        '/api/accident-detection/frame',

        formData

      );


    // ========================================================
    // DEBUG LOG
    // ========================================================

    console.log(

      'ACCIDENT AI RESPONSE:',

      data

    );


    return data as AccidentFrameResponse;


  } catch (error) {


    console.error(

      'Accident AI analysis failed:',

      error

    );


    // Return safe fallback.
    // Never fake accident detection.

    return {

      success: false,

      detections: [],

      accident_detected: false,

      accident_active: false,

      newly_confirmed: false,

      verification: {

        consecutive_hits: 0,

        required_hits: 3,

        missed_frames: 0

      },

      best_accident: null,

      statistics: {

        total_frames: 0,

        confirmed_accidents: 0

      },

      frame_width: 0,

      frame_height: 0

    };

  }

}
// ============================================================
// RESET ACCIDENT DETECTOR
// ============================================================

export async function resetAccidentDetector() {


  try {


    const { data } =
      await api.post(

        '/api/accident-detection/reset'

      );


    console.log(

      'Accident detector reset:',

      data

    );


    return data;


  } catch (error) {


    console.error(

      'Failed to reset accident detector:',

      error

    );


    throw error;

  }

}
// ============================================================
// GET ACCIDENT AI STATUS
// ============================================================

export async function getAccidentAIStatus() {


  const { data } =
    await api.get(

      '/api/accident-detection/status'

    );


  return data;

}